### ISSUE TITLE
RFE: allow missing of type id if target value is concrete

### ISSUE THREAD

**Issue by raderio** [FasterXML/jackson-databind#1716]
```
@JsonTypeInfo( use = JsonTypeInfo.Id.NAME,  include = JsonTypeInfo.As.PROPERTY, property = "type")
@JsonSubTypes({@JsonSubTypes.Type(value = Dog.class, name = "dog"), @JsonSubTypes.Type(value = Cat.class, name = "cat")})
public abstract class Animal {
    private String sound;

    public String getSound() {
        return sound;
    }

    public void setSound(String sound) {
        this.sound = sound;
    }
}

public class Dog extends Animal {
}

public class Cat extends Animal {
}

public static void main(String[] args) throws IOException {
        Animal animal = new ObjectMapper().readValue("{ \"type\": \"cat\", \"sound\": \"mew\" }", Animal.class); // everything ok
        System.out.println(animal.getSound());

        Cat cat = new ObjectMapper().readValue("{ \"sound\": \"mew\" }", Cat.class); // exception
        System.out.println(cat.getSound()); 
    }
}
```
Exception:  
```
com.fasterxml.jackson.databind.JsonMappingException: Unexpected token (END_OBJECT), expected FIELD_NAME: missing property 'type' that is to contain type id  (for class Cat)
 at [Source: { "sound": "mew" };
```
When json is mapped to parent class is ok to send "type" property, but when is mapped to child class, the "type" property must not be required.

**Comment by cowtowncoder** [comment]
This works as expected: type id is required if polymorphic type is enabled and there is no `defaultImpl`.

**Comment by BenDol** [comment]
Is there a way I can get this functioning in the current version? I have some JSON generated by an external mapper that will use type info for some serialization but not for certain types, so my jackson mapper throws an exception:

```
Caused by: com.fasterxml.jackson.databind.exc.InvalidTypeIdException: Missing type id when trying to resolve subtype of [map type; class java.util.Map, [simple type, class java.lang.Object] -> [simple type, class java.lang.Object]]: missing type id property '@class'
```

**Comment by xc350** [comment]
is this issue resolved now?

**Comment by cowtowncoder** [comment]
@xc350 I try to not to be rude here but... what do you think? Open issue on issue tracker tends to indicate that the problem still exists. And this is the case here, to the best of my knowledge.

**Comment by jinmeiliao** [comment]
Can I request this issue to be fixed? If I specifically asked the json string to be parsed into the concrete type `Cat cat = new ObjectMapper().readValue("{ \"sound\": \"mew\" }", Cat.class);`, it would be nice that I don't need to include the type info in the json itself.

**Comment by cowtowncoder** [comment]
@jinmeiliao you do not need to request it: this is an open RFE. It has already been requested.

Being OSS project, whether it will be implemented depends on if anyone with interest and time is willing to implement it.

**Comment by jinmeiliao** [comment]
I actually found a way to achieve this. There is a MapperFeature that already exists for this:
`mapper.configure(MapperFeature.USE_BASE_TYPE_AS_DEFAULT_IMPL, true);`
this will allow `Cat cat = new ObjectMapper().readValue("{ \"sound\": \"mew\" }", Cat.class);` to not throw the exception.

**Comment by christophercolumbusdog** [comment]
@jinmeiliao this is extremely helpful, thank you!

**Comment by nucatus** [comment]
Another option would be to register a _mixin_ for the given target that instructs the unmarshal to ignore the super type and construct the target on its sole context.

> This means that no explicit type metadata is included, and typing is purely done using contextual information possibly augmented with other annotations.

```java
@JsonTypeInfo( use = JsonTypeInfo.Id.NONE )
public class CatMixin {
}

...

mapper.addMixIn(Cat.class, CatMixin.class);
```

**Comment by TolgaDaSilva** [comment]
Any update on this issue?

**Comment by cowtowncoder** [comment]
@TolgaDaSilva No updates.

**Comment by cowtowncoder** [comment]
Fixed as #2968 -- will close this one despite being older because PR was for that issue.
